### PR TITLE
variants of r/w_ok

### PR DESCRIPTION
**Pull request by kroening**
This adds

1) a new predicate `rw_ok`, which is the conjunction of `r_ok` and `w_ok`;

2) variants of `r_ok`, `w_ok` and `rw_ok` that only have one parameter, a pointer.
The size of the memory range that is ok to read or write is given implicitly
by the size of the subtype of the pointer argument.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [X] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: doc/cprover-manual/api.md [hunk 1]
```diff
@@ -191,16 +191,25 @@ option to verify the preconditions of the primitives.
...
-memory starting at the given pointer with the given size is safe.
-**\_\_CPROVER\_w_ok** does the same with writing.
+object.
+
+The function **\_\_CPROVER\_r_ok** returns true if reading the piece of
+memory starting at the given pointer with the given size is safe. 
+**\_\_CPROVER\_w_ok** does the same with writing, and **\_\_CPROVER\_rw_ok**
+returns true when it is safe to do both.  These predicates can be given an
+optional size; when the size is not given, the size of the subtype of the
+pointer type is used.
```
Comment (tautschnig) [inline doc/cprover-manual/api.md]: What is the behaviour when passing a `void *` to the single-argument version? (Reading the code I now know that that's not permitted, but then the docs should spell that out as well.)

**Review by tautschnig — APPROVED** [review]
Looks good to me except for the documentation fix that's still needed.

**Review by peterschrammel — APPROVED** [review]

File: src/ansi-c/c_typecheck_expr.cpp [hunk 2]
```diff
@@ -2558,20 +2558,65 @@ exprt c_typecheck_baset::do_special_functions(
     return std::move(malloc_expr);
   }
   else if(
-    identifier == CPROVER_PREFIX "r_ok" || identifier == CPROVER_PREFIX "w_ok")
+    identifier == CPROVER_PREFIX "r_ok" ||
+    identifier == CPROVER_PREFIX "w_ok" || identifier == CPROVER_PREFIX "rw_ok")
   {
-    if(expr.arguments().size() != 2)
+    if(expr.arguments().size() != 1 && expr.arguments().size() != 2)
```
Comment (NlightNFotis) [inline src/ansi-c/c_typecheck_expr.cpp:2564]: Minor, but this is using our old constraint handling infrastructure - would it be better if this was written as two preconditions? Something like this:

```C
PRECONDITION(expr.arguments().size() != 1);
PRECONDITION(expr.arguments().size() != 2);
```

or

```c
INVARIANT(expr.arguments().size() != 1 || expr.arguments().size() != 2, "r_ok/w_ok/rw_ok expect one or two operands");
```

File: src/ansi-c/cprover_builtin_headers.h [hunk 3]
```diff
@@ -10,8 +10,9 @@ __CPROVER_bool __CPROVER_is_invalid_pointer(const void *);
 _Bool __CPROVER_is_zero_string(const void *);
 __CPROVER_size_t __CPROVER_zero_string_length(const void *);
 __CPROVER_size_t __CPROVER_buffer_size(const void *);
-__CPROVER_bool __CPROVER_r_ok(const void *, __CPROVER_size_t);
-__CPROVER_bool __CPROVER_w_ok(const void *, __CPROVER_size_t);
+__CPROVER_bool __CPROVER_r_ok();
```
Comment (NlightNFotis) [inline src/ansi-c/cprover_builtin_headers.h:13]: Why did these lose their argument types? Am I missing something?

**Review by NlightNFotis — COMMENTED** [review]
Some questions from my part.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_expr.cpp:2564]: No, these aren't preconditions, but checks on user input.

File: src/ansi-c/cprover_builtin_headers.h (hunk 3 above)
Comment (kroening) [reply src/ansi-c/cprover_builtin_headers.h:13]: This is the C syntax for "allow any number of arguments".

**Pull request merged**